lib/jormungandr/test/integration/Cardano/Pool/MetricsSpec.hs

```diff
@@ -0,0 +1,143 @@
+import Cardano.BM.Trace
+    ( nullTracer )
+import Cardano.Pool.Metrics
+    ( monitorStakePools )
+import Cardano.Wallet.Jormungandr
+    ( toSPBlock )
+import Cardano.Wallet.Jormungandr.Launch
+    ( withConfig )
+import Cardano.Wallet.Jormungandr.Network
+    ( JormungandrBackend (..), withJormungandr, withNetworkLayer )
+import Cardano.Wallet.Network
+    ( NetworkLayer, networkTip )
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeRunExceptT )
+import Control.Concurrent
+    ( forkIO, killThread, threadDelay )
+import Control.Concurrent.Async
+    ( race )
+import Control.Exception
+    ( bracket, throwIO )
+import Control.Monad
+    ( void )
+import Data.List
+    ( isPrefixOf )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Text
+    ( Text )
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( pretty )
+import System.FilePath
+    ( (</>) )
+import System.IO.Temp
+    ( withSystemTempDirectory )
+import Test.Hspec
+    ( Spec, around, expectationFailure, it )
+import Test.Integration.Framework.DSL
+    ( eventually )
+
+import qualified Cardano.Pool.DB as Pool
+import qualified Cardano.Pool.DB.Sqlite as Pool
+
+spec :: Spec
+spec = around setup $ do
+    it "a monitorStakePools thread with db, if killed and restarted, catches up\
+        \ with one that stays running" $ \(nl, referenceDB) ->
+        withDB "db.sqlite" $ \db ->
+            withMonitorStakePoolsThread nl db $ \worker -> do
+                eventually $ db `shouldContainSameSlotsAs` referenceDB
+
+                killThread worker
+                waitNBlocks 3 nl
+                -- We assume the (self-)node won't rollback.
+                db `shouldBeBehind` referenceDB
+
+                withMonitorStakePoolsThread nl db $ \_ -> do
+                    eventually $ db `shouldContainSameSlotsAs` referenceDB
+  where
+    shouldBeBehind db ref = do
+        refSlots <- showStoredSlots ref
+        dbSlots <- showStoredSlots db
+        if dbSlots /= refSlots && dbSlots `isPrefixOf` refSlots
+        then return ()
+        else expectationFailure $ mconcat
+            [ "The db with slots:\n    "
+            , show dbSlots
+            , "\nshould be behind the reference db containing:\n    "
+            , show refSlots
+            , "\nbut isn't."
+            ]
+
+    shouldContainSameSlotsAs db ref = do
+        refSlots <- showStoredSlots ref
+        dbSlots <- showStoredSlots db
+        if dbSlots == refSlots
+        then return ()
+        else expectationFailure $ mconcat
+            [ "The db containing slots:\n    "
+            , show dbSlots
+            , "\nshould contain the same slots as the reference db:\n    "
+            , show refSlots
+            , "\nbut they are different."
+            ]
+
+    showStoredSlots :: Pool.DBLayer IO -> IO [Text]
+    showStoredSlots x = map (pretty @_ @Text . slotId)
+        <$> Pool.readPoolProductionCursor x 100
```

Comment:
Arguably a bit weird to 1) use `IO [Text]` and 2) use `100` as `someLargeEnoughValue` to read all slots in the DB.

Reply:
I have now changed it to `readSlots :: Pool.DBLayer IO -> IO [SlotId]` and added a comment about the arbitrary `Int`.